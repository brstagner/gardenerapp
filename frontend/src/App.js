import { useState, useEffect } from 'react';
import Api from './Api';
import Routes from "./Routes";
import './css/app.css';
import { ThemeProvider } from '@mui/material/styles';
import theme from './theme';


function App () {

  const [currUser, setCurrUser] = useState();

  useEffect(() => {
    if ("username" in localStorage && localStorage.username !== undefined) {
      setCurrUser({
        username: localStorage.username,
        userId: localStorage.userId,
        isAdmin: localStorage.isAdmin,
        token: localStorage.token
      });
    }
    else setCurrUser();

  }, []);

  /** Login an existing user */
  async function login (user) {
    try {
      localStorage.clear();
      setCurrUser();
      const res = await Api.loginUser(user);
      setCurrUser({
        username: res.data.username,
        userId: res.data.user_id,
        isAdmin: res.data.is_admin || false,
        token: res.data.token
      });
      localStorage.setItem("username", res.data.username);
      localStorage.setItem("userId", res.data.user_id);
      localStorage.setItem("isAdmin", res.data.is_admin);
      localStorage.setItem("token", res.data.token);
    } catch (errors) {
      console.log(errors);
    }
  }

  /** Register a new user, login that user */
  async function register (user) {
    try {
      localStorage.clear();
      setCurrUser();
      const res = await Api.addUser(user);
      setCurrUser({
        username: res.data.username,
        userId: res.data.user_id,
        isAdmin: res.data.is_admin || false,
        token: res.data.token
      });
      localStorage.setItem("username", res.data.username);
      localStorage.setItem("userId", res.data.user_id);
      localStorage.setItem("isAdmin", res.data.is_admin);
      localStorage.setItem("token", res.data.token);
      return res;
    } catch (errors) {
      console.log(errors);
    }
  }

  /** Update user profile */
  async function editProfile (data) {
    try {
      const res = await Api.editUser(currUser.userId, data, currUser.token);
      console.log(res.data.user);
      localStorage.setItem("username", res.data.user.username);
      localStorage.setItem("userId", res.data.user.user_id);
      localStorage.setItem("isAdmin", res.data.user.is_admin);
      localStorage.setItem("token", res.data.user.token);
      setCurrUser({
        ...currUser,
        username: res.data.user.username,
        userId: res.data.user.user_id,
        isAdmin: res.data.user.is_admin || false
      });
      return "success";
    } catch (error) {
      // setErrors(errors);
      // return `error: ${errors}`;
      console.log(error);
    }
  }

  /** Logout current user, clear state and localStorage */
  const logout = () => {
    localStorage.clear();
    setCurrUser();
  };

  return (
    <div className="App">
      <ThemeProvider theme={theme}>
        <Routes
          currUser={currUser}
          login={login}
          logout={logout}
          register={register}
          editProfile={editProfile}
        />
      </ThemeProvider>
    </div>
  );
}

export default App;
