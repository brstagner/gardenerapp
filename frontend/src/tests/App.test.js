import { render, screen } from '@testing-library/react';
import App from '../App';

const user = { username: "user", userId: "1", isAdmin: "false", token: "user-token" };
const admin = { username: "admin", userId: "2", isAdmin: "true", token: "admin-token" };

// smoke test
it("renders without crashing", function () {
  render(<App currUser={user} />);
  // screen.debug;
});

// snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(<App currUser={user} />);
  expect(asFragment()).toMatchSnapshot();
});