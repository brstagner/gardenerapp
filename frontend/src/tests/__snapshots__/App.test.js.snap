// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="navbar navbar-expand-sm navbar-custom"
    >
      <a
        class="navbar-brand mb-0 h1 fs-2"
        href="/"
      >
        Gardener
      </a>
    </nav>
    <main>
      <div
        class="container"
      >
        <div
          class="container"
        >
          <div
            class="row gy-5"
          >
            <div
              class="col"
            >
              <div
                class="p-3"
              >
                <h4
                  class="d-flex justify-content-center"
                >
                  Have an account?
                </h4>
                <div
                  class="d-flex justify-content-center"
                >
                  <a
                    class="btn btn-primary d-inline-block"
                    href="/login"
                  >
                    Login
                  </a>
                </div>
              </div>
              <div
                class="p-3"
              />
              <div
                class="p-3"
              >
                <h4
                  class="d-flex justify-content-center"
                >
                  New user?
                </h4>
                <div
                  class="d-flex justify-content-center"
                >
                  <a
                    class="btn btn-primary d-inline-block"
                    href="/register"
                  >
                    Register
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
